cpp/src/arrow/type.cc

```diff
@@ -2494,6 +2519,61 @@ const std::vector<std::shared_ptr<DataType>>& PrimitiveTypes() {
   return g_primitive_types;
 }
 
+bool IsBaseBinaryType(std::shared_ptr<DataType> type) {
+  std::call_once(static_data_initialized, InitStaticData);
+  return g_base_binary_types_set.find(type) != g_base_binary_types_set.end();
```

Comment:
This is really an inefficient way to do it instead of simply switching based on the `id()`.